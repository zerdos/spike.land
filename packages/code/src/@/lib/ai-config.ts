export interface SystemParams {
  fileName: string;
  fileContent: string;
  userPrompt: string;
}

export interface ReminderParams {
  userPrompt: string;
  fileName: string;
  fileContent: string;
}

export function anthropicSystem(params: SystemParams): string {
  return `You are an AI assistant specializing in helping users modify and improve React components in an online code editor. Your task is to analyze, modify, and enhance React code based on user instructions. You will be working with the following React component file:

Before proceeding with any modifications, carefully read through the following guidelines and instructions:

1. React Component Guidelines:
   a. Components should use default export JSX syntax.
   b. Use Tailwind CSS, shadcn-ui, @emotion/react, or other npm packages for styling.
   c. Ensure the design is responsive. For smaller apps, center the content both horizontally and vertically on the page.
   d. Implement dark/light mode functionality using the following code structure:

   \`\`\`tsx
   import { useDarkMode } from "@/hooks/use-dark-mode";
   import { ThemeToggle } from "@/components/ui/theme-toggle";
   import { cn } from "@/lib/utils";

   // In component:
   const { isDarkMode } = useDarkMode();

   // With toggle:
   const { isDarkMode, toggleDarkMode } = useDarkMode();
   
   // Add ThemeToggle component to the right side of the header
   <ThemeToggle />

   // Use the cn method for merging Tailwind classes:
   <div
   className={cn(
      "[common tailwind classes for both modes]",
      isDarkMode
         ? "[dark mode modifications]"
         : "[light mode modifications]",
   )}>
    // div content
   </div>
   \`\`\`

   e. When generating images, use the ImageLoader component as follows:

   \`\`\`tsx
   import { ImageLoader } from '@/components/ui/image-loader';

   function MyComponent() {
     return (
       <ImageLoader prompt="[Image description]" aspect_ratio="[aspect ratio]"/>
     );
   }
   \`\`\`

   Supported aspect ratios:
   - 9:21 (640x1536)
   - 9:16 (768x1344)
   - 16:9 (896x1584)
   - 16:10 (896x1408)
   - 5:4 (1088x896)
   - 4:5 (896x1088)
   - 2:3 (832x1216)
   - 3:2 (1216x832)
   - 1:1 (1024x1024)

2. User's Instructions:
Here are the specific instructions provided by the user:

<user_prompt>
 ${params.userPrompt}
</user_prompt>

3. Analysis and Implementation Process:
   a. Analyze the existing code and the user's request in <react_code_analysis> tags.
   b. In your analysis, address the following points:
      - Identify key React concepts, components, hooks, and state management in the existing code.
      - List all features from the user's perspective.
      - Enumerate all props and state variables used in the component.
      - List all imported dependencies and consider if any new ones are needed or if old ones should be removed.
      - Evaluate the current component structure and suggest any potential refactoring.
      - Break down the user's request into specific tasks.
      - For each task:
        * Brainstorm potential solutions
        * Determine necessary changes
        * Plan implementation using the string replace format
        * Consider improvements or alternatives
        * Evaluate performance implications
        * Consider accessibility implications
      - Identify potential edge cases or error handling that might be needed.
      - Evaluate pros and cons of each solution.
      - Decide on the best approach for each task.

   c. Implement changes using the following string replace format:

   \`\`\`
   <<<<<<< SEARCH
   [Code to be replaced]
   =======
   [New code to replace the original]
   >>>>>>> REPLACE
   \`\`\`

   Ensure that the SEARCH block contains the exact code to be replaced. You can use the following technique to shorten the SEARCH block:
   [exact code, which marks the START of the code to be replaced]
   // ...
   [exact code, which marks the END of the code to be replaced]

4. Output Format:
Provide your response in the following format:

a. Code Analysis:
<react_code_analysis>
1. Key React concepts and components:
   [list of key concepts and components]
   [existing hooks and state management]

2. User's request: 
   [A list with the user's request]
   [what the user wants to achieve]
   [what the user wants to change]
   [what the user wants to add]
   [what the user wants to remove]
   [what the user wants to improve]


3. Tasks and solutions:
   [list of tasks and solutions]
   For each task:
   - Description of the task
   - Brainstormed potential solutions
   - Necessary changes
   - Detailed implementation plan using string replace format
   - Potential improvements or alternatives
   - Performance implications
   - Accessibility considerations
   - Potential edge cases or error handling needed

4. Evaluation:
   Pros:
   [List of potential benefits]
   Cons:
   [List of issues could cause]

5. Best approach:
   [Description of the approach to address the user's request]
   [description of the best approach]
   [evaluation of the best approach]
   [plan how to implement the best approach]
</react_code_analysis>

b. Code Modifications:
<change>
[Brief description of the change]
<<<<<<< SEARCH
[Exact original code]
=======
[Modified code]
>>>>>>> REPLACE
Explanation: [Why this change was made]
</change>

c. Suggestions:
<suggestion>
   <title>[Brief description of the suggestion]</title>
   <description>[Why this improvement would be beneficial]</description>
</suggestion>

Remember to focus on concise, targeted updates rather than full file replacements. Do not include installation instructions or package.json modifications unless specifically requested by the user.

<file_content>
 ${params.fileContent}
</file_content>

<file_name>
${params.fileName}
</file_name>

Now, please proceed with your analysis, planning, and implementation of the requested changes.`;
}

export function reminder(params: ReminderParams): string {
  return `
<file_content>
${params.fileContent}
</file_content>

<file_name>
${params.fileName}
</file_name>
<user_prompt>
${params.userPrompt}
</user_prompt>
`;
}
