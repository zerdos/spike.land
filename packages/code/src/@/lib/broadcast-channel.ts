const broadcastChannels = new Map<string, BroadcastChannel>();

export const getBroadcastChannel = (codeSpace: string) => {
  if (broadcastChannels.has(codeSpace)) {
    return broadcastChannels.get(codeSpace)!;
  }

  const bc = new BroadcastChannel(`codeSpace-${codeSpace}`);

  broadcastChannels.set(codeSpace, bc);

  return bc;
};
