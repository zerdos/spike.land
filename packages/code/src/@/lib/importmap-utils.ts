// src/importMapUtils.ts

export interface ImportMap {
  imports: Record<string, string>;
}

// Configuration Constants
const FILE_EXTENSIONS = new Set([
  ".js",
  ".mjs",
  ".ts",
  ".tsx",
  ".jsx",
  ".json",
  ".wasm",
  ".txt",
  ".svg",
  ".md",
  ".html",
  ".css",
  ".scss",
  ".sass",
  ".less",
  ".styl",
  ".graphql",
  ".gql",
  ".yml",
  ".toml",
  ".xml",
  ".csv",
  ".tsv",
  ".ini",
  ".properties",
  ".env",
  ".env.local",
  ".env.development",
  ".env.test",
  ".env.production",
  ".env.staging",
]);

const WORKER_PATTERNS = ["/workers/", ".worker"];
const COMPONENT_PATTERNS = [
  "@/components/",
  "@/services/",
  "@/config/",
  "@/utils/",
  "@/tools/",
  "@/workflows/",
  "/live/",
  "@/lib",
  "@/external",
  "@/hooks",
];

const EXTERNAL_DEPENDENCIES = [
  "react",
  "react-dom",
  "framer-motion",
  "@emotion/react",
  "@emotion/styled",
];

// Default Import Map
export const importMap: ImportMap = {
  imports: {
    "/@/": "/@/",
    "var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
import { jsx, jsxs } from "@emotion/react/jsx-runtime";
var stdin_default = /* @__PURE__ */ __name(() => /* @__PURE__ */ jsxs("div", { children: [
  /* @__PURE__ */ jsx("h1", { children: "404 - for now." }),
  /* @__PURE__ */ jsx("h2", { children: "But you can edit even this page and share with your friends." })
] }), "default");
export {
  stdin_default as default
};
"
    "@emotion/react/jsx-runtime": "/emotionJsxRuntime.mjs",
    "@emotion/react/jsx-dev-runtime": "/emotionJsxRuntime.mjs",
    "@emotion/styled": "/emotionStyled.mjs",
    "react/jsx-runtime": "/jsx.mjs",
    "react-dom/server": "/reactDomServer.mjs",
    "react-dom/client": "/reactDomClient.mjs",
    "@emotion/react": "/emotion.mjs",
    "react": "/reactMod.mjs",
    "framer-motion": "/@/workers/framer-motion.mjs",
    "react-dom": "/reactDom.mjs",
  },
};

// Utility Functions
function hasKnownExtension(path: string): boolean {
  const lastDotIndex = path.lastIndexOf(".");
  return lastDotIndex !== -1 && FILE_EXTENSIONS.has(path.slice(lastDotIndex));
}

function isWorkerFile(path: string): boolean {
  return WORKER_PATTERNS.some(pattern => path.includes(pattern));
}

function isComponentFile(path: string): boolean {
  return COMPONENT_PATTERNS.some(pattern => path.includes(pattern));
}

function processQueryAndHash(path: string): {
  basePath: string;
  query: string;
  hash: string;
} {
  const [pathPart, hash] = path.split("#");
  const [basePath, query] = pathPart.split("?");

  return {
    basePath: basePath || "",
    query: query ? `?${query}` : "",
    hash: hash ? `#${hash}` : "",
  };
}

function shouldTransformPath(path: string): boolean {
  return (
    !path.includes("?bundle=true") &&
    !path.startsWith("data:") &&
    !path.startsWith("http://") &&
    !path.startsWith("https://") &&
    !path.startsWith("/live/")
  );
}

function getExportsString(match: string): string {
  const namedImports = match.match(/\{([^}]*)\}/s)?.[1];
  return namedImports
    ?.split(",")
    .map((s) => s.trim().split(" as ")[0])
    .filter(Boolean)
    .join(",") || "";
}

function getMappedPath(
  path: string,
  exportsParam = "",
  hasFromClause = false,
  importMapImports: ImportMap["imports"] = importMap.imports,
): string {
  const { basePath, query, hash } = processQueryAndHash(path);

  // Early returns for complex or special paths
  if (path.includes("${") || path.includes("+")) return path;

  // Exact match in import map
  if (importMapImports[basePath]) return path;

  // Handle worker files
  if (isWorkerFile(basePath)) {
    const baseWithoutExt = hasKnownExtension(basePath)
      ? basePath.substring(0, basePath.lastIndexOf("."))
      : basePath;
    const extension = hasFromClause ? ".mjs" : ".js";
    const resultPath = baseWithoutExt + extension;

    return !resultPath.startsWith(".") && !resultPath.startsWith("/")
      ? `/${resultPath}`
      : resultPath;
  }

  // Handle component and special files
  if (isComponentFile(basePath)) {
    const baseWithoutExt = hasKnownExtension(basePath)
      ? basePath.substring(0, basePath.lastIndexOf("."))
      : basePath;
    const extension = ".mjs";
    const resultPath = baseWithoutExt + extension;

    return !resultPath.startsWith(".") && !resultPath.startsWith("/")
      ? `/${resultPath}`
      : resultPath;
  }

  // Handle directory imports and paths without extension
  if (
    !hasKnownExtension(basePath) &&
    (basePath.startsWith(".") || basePath.startsWith("/"))
  ) {
    const extension = basePath.endsWith("/") ? "index.mjs" : ".mjs";
    return `${basePath}${extension}${query}${hash}`;
  }

  // Preserve live and absolute URLs
  if (basePath.startsWith("/live/") || /^https?:\/\//.test(basePath)) {
    return path;
  }

  // Handle non-relative paths
  if (!basePath.startsWith(".") && !basePath.startsWith("/")) {
    const params = [
      "bundle=true",
      `external=${EXTERNAL_DEPENDENCIES.join(",")}`,
    ];

    if (exportsParam) params.push(`exports=${exportsParam}`);

    return `/${basePath}?${params.join("&")}${query}${hash}`;
  }

  return `${basePath}${query}${hash}`;
}

export function importMapReplace(
  code: string,
): string {
  // Prevent double processing
  if (code.includes("/** importMapReplace */") || code.includes("/* esm.sh")) {
    return code;
  }

  // Normalize line endings
  code = code.replace(/\r\n/g, "\n");

  // Helper function for replacing imports
  const replaceImport = (match: string, pre: string, q1: string, path: string, q2: string) => {
    if (!shouldTransformPath(path)) return match;
    const fullPath = getMappedPath(path, "", false);
    return `${pre}${q1}${fullPath}${q2}`;
  };

  // Replace simple imports
  code = code.replace(
    /^(\s*import\s+)(['"])([^'"]+)(['"];\s*$)/gm,
    replaceImport,
  );

  // Replace imports with from clause
  code = code.replace(
    /^(\s*import\s+[\s\S]*?from\s+['"])([^'"]+)(['"];?)/gm,
    (match) => {
      if (match.includes("/** importMapReplace */")) return match;

      const pathMatch = match.match(/['"]([^'"]+)['"]/);
      if (!pathMatch) return match;

      const importPath = pathMatch[1];
      if (!shouldTransformPath(importPath)) return match;

      const exportsParam = getExportsString(match);
      const mappedPath = getMappedPath(importPath, exportsParam, true);
      return match.replace(/['"][^'"]+['"]/, `"${mappedPath}"`);
    },
  );

  // Replace export from clauses
  code = code.replace(
    /^(\s*export\s+[\s\S]*?from\s+['"])([^'"]+)(['"];?)/gm,
    (match) => {
      if (match.includes("/** importMapReplace */")) return match;

      const pathMatch = match.match(/['"]([^'"]+)['"]/);
      if (!pathMatch) return match;

      const exportPath = pathMatch[1];
      if (!shouldTransformPath(exportPath)) return match;

      const exportsParam = getExportsString(match);
      const mappedPath = getMappedPath(exportPath, exportsParam, true);
      return match.replace(/['"][^'"]+['"]/, `"${mappedPath}"`);
    },
  );

  // Replace dynamic imports
  code = code.replace(
    /(\bimport\s*\(\s*['"])([^'"]+)(['"]\s*\))/g,
    (match) => {
      const pathMatch = match.match(/['"]([^'"]+)['"]/);
      if (!pathMatch) return match;

      const importPath = pathMatch[1];
      if (!shouldTransformPath(importPath)) return match;

      const mappedPath = getMappedPath(importPath, "", true);
      return match.replace(/['"][^'"]+['"]/, `"${mappedPath}"`);
    },
  );

  return `/** importMapReplace */\n${code}`;
}

export default importMap;