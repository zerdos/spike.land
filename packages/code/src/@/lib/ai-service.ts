import { anthropicSystem, reminder } from "@/lib/ai-config";
import type { Message } from "@/lib/interfaces";
import Anthropic from "@anthropic-ai/sdk";
import OpenAI from "openai";

export interface AIServiceConfig {
  anthropicEndpoint: string;
  openAIEndpoint: string;
  gpt4oEndpoint: string;
  updateThrottleMs: number;
  retryWithClaudeEnabled: boolean;
  setIsStreaming: (isStreaming: boolean) => void;
}

export class AIService {
  private _config: AIServiceConfig;

  constructor(config: AIServiceConfig) {
    this._config = config;
  }

  public async sendToAnthropic(
    messages: Message[],
    onUpdate: (code: string) => void,
  ): Promise<Message> {
    this._config.setIsStreaming(true);

    try {
      const client = new Anthropic({
        dangerouslyAllowBrowser: true,
        apiKey: 'DUMMY_API_KEY',
        baseURL: this._config.anthropicEndpoint,
      });

      const stream = await client.messages.create({
        max_tokens: 8192,
        messages: messages.map((message) => ({
          role: message.role === "system" ? "user" : message.role,
          content: message.content.toString(),
        })),
        model: "claude-3-5-sonnet-latest",
        stream: true,
      });

      let content = "";
      for await (const messageStreamEvent of stream) {
        if (
          "type" in messageStreamEvent &&
          messageStreamEvent.type === "content_block_delta" &&
          messageStreamEvent.delta.type === "text_delta"
        ) {
          const chunkContent = messageStreamEvent.delta.text;
          if (chunkContent) {
            content += chunkContent;
            onUpdate(chunkContent);
          }
       
        }
      }

      return {
        id: Math.random().toString(36).substring(7),
        role: "assistant",
        content,
      };
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error("Unknown error occurred");
    } finally {
      this._config.setIsStreaming(false);
    }
  }

  public async sendToGpt4o(
    messages: Message[],
    onUpdate: (code: string) => void,
  ): Promise<Message> {
    this._config.setIsStreaming(true);

    try {
      const client = new OpenAI({
        baseURL: this._config.gpt4oEndpoint,
        apiKey: 'DUMMY_API_KEY',
      });

      const stream = await client.chat.completions.create({
        model: "gpt-4o-mini",
        stream: true,
        messages: messages.map((message) => ({
          role: message.role === "system" ? "user" : message.role,
          content: message.content.toString(),
        })),
      });

      let content = "";
      for await (const chunk of stream) {
        const chunkContent = chunk.choices[0]?.delta?.content;
        if (chunkContent) {
          content += chunkContent;
          onUpdate(chunkContent);
        }
      }

      return {
        id: Math.random().toString(36).substring(7),
        role: "assistant",
        content,
      };
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error("Unknown error occurred");
    } finally {
      this._config.setIsStreaming(false);
    }
  }

  public prepareClaudeContent(
    params: { content: string; messages: Message[]; codeNow: string; codeSpace: string; },
  ): string {
    const { content, messages, codeNow, codeSpace } = params;

    // Check for error messages that should be passed through directly
    if (content.toLowerCase().startsWith("i'm sorry")) {
      return content;
    }

    const lastMessage = messages[messages.length - 1];

    // If codeNow is different from the last message or there are no messages,
    // use anthropic system content
    // if (!lastMessage || lastMessage.content !== codeNow) {
      const result = anthropicSystem({
        fileName: codeSpace+".tsx",
        fileContent: codeNow,
        userPrompt: content,
      });
      return result;
    // }

    // If code hasn't changed, return reminder content
    // return reminder({
    //   userPrompt: content,
    //   fileName: codeSpace+ ".tsx",
    //   fileContent: codeNow,
    // });
  }
}
