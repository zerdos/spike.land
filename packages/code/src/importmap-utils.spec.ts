import { importMapReplace } from "@/lib/importmap-utils";
import { describe, expect, it } from "vitest";

describe("importMapReplace", () => {
  const origin = "http://localhost:3000";

  const scenarios = {
    // Basic imports
    "basic named imports": `import { prop, prop2 } from "foo";`,
    "specific exports": `import { __await, __rest } from "tslib";`,
    "specific named imports with aliases":
      `import { __await as aw, __rest as restNow} from "tslib";`,
    "default import": 
      `import React from "react";`,
    "default and named imports together":
      `import React, { useState, useEffect } from "react";`,
    "multiple named imports with complex names":
      `import { camelCase, snake_case, PascalCase, CONSTANT_CASE } from "utils";`,
    
    // Worker and special imports
    "should transpile worker files to js with wildcard import":
      `import * as Monaco from "@/workers/monaco-editor.worker";`,
    "should transpile worker files to js with bare import":
      `import "@/workers/monaco-editor.worker";`,
    "service worker registration":
      `import { register } from "@/workers/service-worker";`,
    "web worker imports":
      `import Worker from "@/workers/computation.worker";`,
      
    // Path handling
    "relative imports without extension": 
      `import { helper } from "./utils/helper";`,
    "parent directory relative imports":
      `import { shared } from "../shared/utils";`,
    "should preserve existing extensions in relative imports":
      `import { component } from "./components/Button";`,
    "deep nested paths":
      `import { util } from "../../../../very/deep/nested/path";`,
    "current directory marker":
      `import { tool } from "./././current/path";`,
    
    // URL types
    "data URLs":
      `import { foo } from "data:text/javascript,export const foo = 'bar'";`,
    "live URLs":
      `import { Component } from "/live/components/Button";`,
    "http URLs from same origin":
      `import { lib } from "http://localhost:3000/lib?bundle=true&exports=foo";`,
    "http URLs from different origin":
      `import { lib } from "http://example.com/lib?bundle=true&exports=foo";`,
    "https URLs":
      `import { secure } from "https://example.com/secure-module";`,
    "URLs with complex query parameters":
      `import { module } from "https://example.com/module?version=1.2.3&format=esm&debug=true";`,
    
    // Dynamic imports
    "dynamic imports":
      `const mod = await import("module");`,
    "dynamic imports with template literals":
      `const mod = await import(\`./modules/\${name}\`);`,
    "dynamic imports with complex expressions":
      `const mod = await import(process.env.NODE_ENV === 'production' ? 'prod' : 'dev');`,
    "dynamic imports with relative paths":
      `const mod = await import('./relative/path/module');`,
    
    // Export variations
    "import with export":
      `export { foo } from "bar";`,
    "export with rename":
      `export { foo as default } from "module";`,
    "multiple exports":
      `export { foo, bar as default, baz as customName } from "module";`,
    "export with destructuring":
      `export const { foo, bar } = await import("module");`,
    
    // Special cases and edge cases
    "should prevent double processing":
      `/** importMapReplace */
    import { foo } from "bar";`,
    "imports with comments":
      `// This is a comment
    import { foo } from "bar"; // End of line comment`,
    "multi-line imports":
      `import {
      foo,
      bar,
      baz
    } from "module";`,
    "imports with unusual whitespace":
      `import    {    foo   ,   bar   }    from     "module"   ;`,
    "imports with unicode characters":
      `import { π, λ, γ } from "math-symbols";`,
    "import with empty specifiers":
      `import {} from "empty-module";`,
    
    // Module types
    "CSS modules":
      `import styles from "./styles.module.css";`,
    "JSON modules":
      `import data from "./data.json";`,
    "WASM modules":
      `import * as wasm from "./module.wasm";`,
    "TypeScript path aliases":
      `import { component } from "@components/Button";`,
    
    // URL parameters and fragments
    "URLs with hash fragments":
      `import { section } from "module#section";`,
    "URLs with version tags":
      `import { lib } from "https://esm.sh/lodash@4.17.21";`,
    "scoped packages":
      `import { component } from "@scope/package/component";`,
  };

  for (const [description, code] of Object.entries(scenarios)) {
    it("should handle " +description, async () => {
      const result = importMapReplace(code, origin);
      expect({result, code}).toMatchSnapshot();
    });
  }

});
