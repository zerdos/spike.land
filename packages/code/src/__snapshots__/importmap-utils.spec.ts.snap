// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`importMapReplace > should handle CSS modules 1`] = `
{
  "code": "import styles from "./styles.module.css";",
  "result": "/** importMapReplace */
import styles from "./styles.module.css.mjs";",
}
`;

exports[`importMapReplace > should handle JSON modules 1`] = `
{
  "code": "import data from "./data.json";",
  "result": "/** importMapReplace */
import data from "./data.json.mjs";",
}
`;

exports[`importMapReplace > should handle TypeScript path aliases 1`] = `
{
  "code": "import { component } from "@components/Button";",
  "result": "/** importMapReplace */
import { component } from "http://localhost:3000/@components/Button?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=component";",
}
`;

exports[`importMapReplace > should handle URLs with complex query parameters 1`] = `
{
  "code": "import { module } from "https://example.com/module?version=1.2.3&format=esm&debug=true";",
  "result": "/** importMapReplace */
import { module } from "https://example.com/module?version=1.2.3&format=esm&debug=true";",
}
`;

exports[`importMapReplace > should handle URLs with hash fragments 1`] = `
{
  "code": "import { section } from "module#section";",
  "result": "/** importMapReplace */
import { section } from "http://localhost:3000/module#section?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=section";",
}
`;

exports[`importMapReplace > should handle URLs with version tags 1`] = `
{
  "code": "import { lib } from "https://esm.sh/lodash@4.17.21";",
  "result": "/** importMapReplace */
import { lib } from "https://esm.sh/lodash@4.17.21";",
}
`;

exports[`importMapReplace > should handle WASM modules 1`] = `
{
  "code": "import * as wasm from "./module.wasm";",
  "result": "/** importMapReplace */
import * as wasm from "./module.wasm.mjs";",
}
`;

exports[`importMapReplace > should handle basic named imports 1`] = `
{
  "code": "import { prop, prop2 } from "foo";",
  "result": "/** importMapReplace */
import { prop, prop2 } from "http://localhost:3000/foo?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=prop,prop2";",
}
`;

exports[`importMapReplace > should handle current directory marker 1`] = `
{
  "code": "import { tool } from "./././current/path";",
  "result": "/** importMapReplace */
import { tool } from "./././current/path.mjs";",
}
`;

exports[`importMapReplace > should handle data URLs 1`] = `
{
  "code": "import { foo } from "data:text/javascript,export const foo = 'bar'";",
  "result": "/** importMapReplace */
import { foo } from "data:text/javascript,export const foo = /index.js"bar'";",
}
`;

exports[`importMapReplace > should handle deep nested paths 1`] = `
{
  "code": "import { util } from "../../../../very/deep/nested/path";",
  "result": "/** importMapReplace */
import { util } from "../../../../very/deep/nested/path.mjs";",
}
`;

exports[`importMapReplace > should handle default and named imports together 1`] = `
{
  "code": "import React, { useState, useEffect } from "react";",
  "result": "/** importMapReplace */
import React, { useState, useEffect } from "react";",
}
`;

exports[`importMapReplace > should handle default import 1`] = `
{
  "code": "import React from "react";",
  "result": "/** importMapReplace */
import React from "react";",
}
`;

exports[`importMapReplace > should handle dynamic imports 1`] = `
{
  "code": "const mod = await import("module");",
  "result": "/** importMapReplace */
const mod = await import("http://localhost:3000/module?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled");",
}
`;

exports[`importMapReplace > should handle dynamic imports with complex expressions 1`] = `
{
  "code": "const mod = await import(process.env.NODE_ENV === 'production' ? 'prod' : 'dev');",
  "result": "/** importMapReplace */
const mod = await import(process.env.NODE_ENV === 'production' ? 'prod' : 'dev');",
}
`;

exports[`importMapReplace > should handle dynamic imports with relative paths 1`] = `
{
  "code": "const mod = await import('./relative/path/module');",
  "result": "/** importMapReplace */
const mod = await import("relative/path/module.mjs");",
}
`;

exports[`importMapReplace > should handle dynamic imports with template literals 1`] = `
{
  "code": "const mod = await import(\`./modules/\${name}\`);",
  "result": "/** importMapReplace */
const mod = await import(\`./modules/\${name}\`);",
}
`;

exports[`importMapReplace > should handle export with destructuring 1`] = `
{
  "code": "export const { foo, bar } = await import("module");",
  "result": "/** importMapReplace */
export const { foo, bar } = await import("http://localhost:3000/module?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled");",
}
`;

exports[`importMapReplace > should handle export with rename 1`] = `
{
  "code": "export { foo as default } from "module";",
  "result": "/** importMapReplace */
export { foo as default } from "http://localhost:3000/module?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled";",
}
`;

exports[`importMapReplace > should handle http URLs from different origin 1`] = `
{
  "code": "import { lib } from "http://example.com/lib?bundle=true&exports=foo";",
  "result": "/** importMapReplace */
import { lib } from "http://example.com/lib?bundle=true&exports=foo";",
}
`;

exports[`importMapReplace > should handle http URLs from same origin 1`] = `
{
  "code": "import { lib } from "http://localhost:3000/lib?bundle=true&exports=foo";",
  "result": "/** importMapReplace */
import { lib } from "http://localhost:3000/lib?bundle=true&exports=foo";",
}
`;

exports[`importMapReplace > should handle https URLs 1`] = `
{
  "code": "import { secure } from "https://example.com/secure-module";",
  "result": "/** importMapReplace */
import { secure } from "https://example.com/secure-module";",
}
`;

exports[`importMapReplace > should handle import with empty specifiers 1`] = `
{
  "code": "import {} from "empty-module";",
  "result": "/** importMapReplace */
import {} from "http://localhost:3000/empty-module?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled";",
}
`;

exports[`importMapReplace > should handle import with export 1`] = `
{
  "code": "export { foo } from "bar";",
  "result": "/** importMapReplace */
export { foo } from "http://localhost:3000/bar?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled";",
}
`;

exports[`importMapReplace > should handle imports with comments 1`] = `
{
  "code": "// This is a comment
    import { foo } from "bar"; // End of line comment",
  "result": "/** importMapReplace */
import { foo } from "http://localhost:3000/bar?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=foo"; // End of line comment",
}
`;

exports[`importMapReplace > should handle imports with unicode characters 1`] = `
{
  "code": "import { π, λ, γ } from "math-symbols";",
  "result": "/** importMapReplace */
import { π, λ, γ } from "math-symbols";",
}
`;

exports[`importMapReplace > should handle imports with unusual whitespace 1`] = `
{
  "code": "import    {    foo   ,   bar   }    from     "module"   ;",
  "result": "/** importMapReplace */
import    {    foo   ,   bar   }    from     "http://localhost:3000/module?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=foo,bar"   ;",
}
`;

exports[`importMapReplace > should handle live URLs 1`] = `
{
  "code": "import { Component } from "/live/components/Button";",
  "result": "/** importMapReplace */
import { Component } from "/live/components/Button/index.js";",
}
`;

exports[`importMapReplace > should handle multi-line imports 1`] = `
{
  "code": "import {
      foo,
      bar,
      baz
    } from "module";",
  "result": "/** importMapReplace */
import {
foo,
bar,
baz
} from "http://localhost:3000/module?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=foo,bar,baz";",
}
`;

exports[`importMapReplace > should handle multiple exports 1`] = `
{
  "code": "export { foo, bar as default, baz as customName } from "module";",
  "result": "/** importMapReplace */
export { foo, bar as default, baz as customName } from "http://localhost:3000/module?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled";",
}
`;

exports[`importMapReplace > should handle multiple named imports with complex names 1`] = `
{
  "code": "import { camelCase, snake_case, PascalCase, CONSTANT_CASE } from "utils";",
  "result": "/** importMapReplace */
import { camelCase, snake_case, PascalCase, CONSTANT_CASE } from "http://localhost:3000/utils?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=camelCase,snake_case,PascalCase,CONSTANT_CASE";",
}
`;

exports[`importMapReplace > should handle parent directory relative imports 1`] = `
{
  "code": "import { shared } from "../shared/utils";",
  "result": "/** importMapReplace */
import { shared } from "../shared/utils.mjs";",
}
`;

exports[`importMapReplace > should handle relative imports without extension 1`] = `
{
  "code": "import { helper } from "./utils/helper";",
  "result": "/** importMapReplace */
import { helper } from "utils/helper.mjs";",
}
`;

exports[`importMapReplace > should handle scoped packages 1`] = `
{
  "code": "import { component } from "@scope/package/component";",
  "result": "/** importMapReplace */
import { component } from "http://localhost:3000/@scope/package/component?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=component";",
}
`;

exports[`importMapReplace > should handle service worker registration 1`] = `
{
  "code": "import { register } from "@/workers/service-worker";",
  "result": "/** importMapReplace */
import { register } from "/@/workers/service-worker.mjs";",
}
`;

exports[`importMapReplace > should handle should preserve existing extensions in relative imports 1`] = `
{
  "code": "import { component } from "./components/Button";",
  "result": "/** importMapReplace */
import { component } from "components/Button.mjs";",
}
`;

exports[`importMapReplace > should handle should prevent double processing 1`] = `
{
  "code": "/** importMapReplace */
    import { foo } from "bar";",
  "result": "/** importMapReplace */
    import { foo } from "bar";",
}
`;

exports[`importMapReplace > should handle should transpile worker files to js with bare import 1`] = `
{
  "code": "import "@/workers/monaco-editor.worker";",
  "result": "/** importMapReplace */
import "http://localhost:3000/@/workers/monaco-editor.worker.js";",
}
`;

exports[`importMapReplace > should handle should transpile worker files to js with wildcard import 1`] = `
{
  "code": "import * as Monaco from "@/workers/monaco-editor.worker";",
  "result": "/** importMapReplace */
import * as Monaco from "/@/workers/monaco-editor.worker.mjs";",
}
`;

exports[`importMapReplace > should handle specific exports 1`] = `
{
  "code": "import { __await, __rest } from "tslib";",
  "result": "/** importMapReplace */
import { __await, __rest } from "http://localhost:3000/tslib?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=__await,__rest";",
}
`;

exports[`importMapReplace > should handle specific named imports with aliases 1`] = `
{
  "code": "import { __await as aw, __rest as restNow} from "tslib";",
  "result": "/** importMapReplace */
import { __await as aw, __rest as restNow} from "http://localhost:3000/tslib?bundle=true&external=react,react-dom,framer-motion,@emotion/react,@emotion/styled&exports=__await,__rest";",
}
`;

exports[`importMapReplace > should handle web worker imports 1`] = `
{
  "code": "import Worker from "@/workers/computation.worker";",
  "result": "/** importMapReplace */
import Worker from "/@/workers/computation.worker.mjs";",
}
`;
